numba/cuda/cudaimpl.py

```diff
@@ -492,6 +492,141 @@ def ptx_fp16_binary(context, builder, sig, args):
+    cvt2fp16_asm = ir.InlineAsm(cvt2fp16_fnty,
+                                "cvt.rn.f16.f32 $0, $1;",
+                                "=h,f")
+    fp16_div = builder.call(cvt2fp16_asm, [fmul])
+
+    # Before performing a second Newton Rapson iteration
+    # filter out small denormal values and zero.
+    # If either case is true skip the second iteration.
+
+    and_cnst = context.get_constant(types.int16, 32767)
+    and_op = builder.and_(fp16_div, and_cnst)
+
+    # Check for small denormals
+    temp_reg = '__$$temp3'
+    reg_pred_fnty = ir.FunctionType(ir.VoidType(),[])
+    reg_pred_asm = ir.InlineAsm(reg_pred_fnty,
+                                f"{{ .reg .pred {temp_reg};",
+                                "")
+    _ = builder.call(reg_pred_asm, [])
+
+    setp_fnty = ir.FunctionType(ir.VoidType(),
+                                [ir.IntType(16), ir.IntType(16)])
+    setp_cnst = context.get_constant(types.int16, 143)
+    setp_asm = ir.InlineAsm(setp_fnty,
+                            f"setp.lt.f16 {temp_reg}, $0, $1;",
+                            'h,h')
+    _ = builder.call(setp_asm, [and_op, setp_cnst])
+
+    selp_fnty = ir.FunctionType(ir.IntType(16),[])
+
+    selp_asm = ir.InlineAsm(selp_fnty,
+                            f"selp.u16 $0, 1, 0, {temp_reg};}}",
+                            '=h')
+    selp = builder.call(selp_asm, [])
+
+    zero = context.get_constant(types.int16, 0)
+    cmp1 = builder.icmp_unsigned("==",
+                                 builder.bitcast(selp, ir.IntType(16)),
+                                 zero)
+
+    # Check for zero result
+    cmp2 = builder.icmp_unsigned("==",
+                                 builder.bitcast(and_op, ir.IntType(16)),
+                                 zero)
+    # Filter out denormal and precise zero
+    or_op = builder.or_(cmp1, cmp2)
+    or_cond = builder.bitcast(or_op, ir.IntType(1))
+
+    entry_bb = builder.block
+    fall_thru_bb = builder.append_basic_block()
+    true_bb = builder.append_basic_block()
+    builder.cbranch(or_cond, true_bb, fall_thru_bb)
+
+    # Fall through branch (Second Newton Rapson iteration)
+    builder.position_at_start(fall_thru_bb)
+
+    float_zero = context.get_constant(types.float32, 0.0)
+    neg_f = builder.fsub(float_zero, arg2_fp32)
+
+    fname = 'llvm.nvvm.fma.rn.f'
+    lmod = builder.module
+    fnty = ir.FunctionType(ir.FloatType(),
+                           (ir.FloatType(),
+                            ir.FloatType(),
+                            ir.FloatType()))
+    fma_func = cgutils.get_or_insert_function(lmod, fnty, fname)
+
+    # Mathematically we have: q0 = x * 1/y * (1 - err)
+    # So err (correction_term) = x - y*q0 = fma(-y, q0, x)
+    # And q1 = err*yn + q0 = fma(err, yn, q0)
+
+    fma1 = builder.call(fma_func, [neg_f, fmul, arg2_fp32])
+    fma2 = builder.call(fma_func, [arg2_rcp, fma1, fmul])
+    fp32_to_f16 = builder.call(cvt2fp16_asm, [fma2])
+    builder.branch(true_bb)
```

Comment:
This should be moved to the next block of code (after the comment, that is).

Reply:
@isVoid I don't understand this comment.

Reply:
Apologies, I wasn't aware this is the unconditional jump to the true branch. Then this line should stay where it is.